### ISSUE TITLE
Add `LOCAL_ONLY` extension to `SELECT`, `INSERT` and `UPDATE` statements

### ISSUE THREAD

**Issue by denesb** [scylladb/scylladb#16478]
A common task when writing tests, is to control which nodes certain writes and up on, then verifying the data that individual nodes hold. Currently, this is painful, full of pitfalls and leads to slow tests.

## Writes

The underlying problem is that writes, regardless of `CL`, will be sent to all nodes in the cluster, the only thing `CL` determines is the number of responses to wait for before responding to the client. So the usual way to control which nodes receive a write, is to stop/kill all other nodes in a cluster. This is slow even in a 2-node cluster, as one has to wait for all other nodes to notice that the node was stopped/started, on each operation (otherwise the coordinator will still attempt to send the write and once the RPC connection is reestablished, it might succeed, if it didn't time out yet).
One also has to make sure to disable HH, read-repair and so on.

An alternative solution (used in a single test so far) is to generate sstables with `scylla sstable write` and upload them to nodes. This is much more painful to code and doesn't help much with the runtime of the test.

## Reads

Reads are a bit easier to control. If great care was taken on the client side, when setting up the connection, single partition reads with `CL=ONE` will reads from just one node. This only works for single-partition reads, int he case of range-scans, there is no way to control replica selection at all. So the options are: stop all other nodes, or use `SELECT * FROM MUTATION_FRAGMENTS()`, which is *very* clunky to use, when one just wants to read the user-visible data (basically requires implementing combined reader + compacting reader on client side).

## Proposed solution

Implement an extension, similar to the already existing `SELECT * FROM MUTATION_FRAGMENTS(table)`. This extension would be called `LOCAL_ONLY`, and would modify the statement such that it reads/writes strictly from/to the local node. Switching between coordinators (load-balancing) would be prohibited, just like it is for `MUTATION_FRAGMENTS()` extension.

**Comment by denesb** [comment]
cc @bhalevy @nyh @avikivity

**Comment by denesb** [comment]
Linking here the PR which introduced the  `SELECT * FROM MUTATION_FRAGMENTS()`: https://github.com/scylladb/scylladb/pull/14347

**Comment by nyh** [comment]
This issue is a duplicate of #9042 - although probably we should keep this one and close the old one because this one offers a more concrete plan. One thing mentioned in #9042 and not yet mentioned in this issue is that another complication for reads before adding this feature was HWLB - which may decide to read from another node even if this one did have the data.

We need to decide (and document) what LOCAL_ONLY does when the local node is _not_ an owner of the requested partition. I think for reads we should just return what data is available locally or nothing - this will also allow a full table scan to return all the local data. For writes I think it should be an error to attempt to write to a partition not owned by this node.

**Comment by denesb** [comment]
> We need to decide (and document) what LOCAL_ONLY does when the local node is _not_ an owner of the requested partition. I think for reads we should just return what data is available locally or nothing - this will also allow a full table scan to return all the local data. For writes I think it should be an error to attempt to write to a partition not owned by this node.

I agree, it should return whatever the node has locally, without considering token ownership.

**Comment by denesb** [comment]
> This issue is a duplicate of #9042 - although probably we should keep this one and close the old one because this one offers a more concrete plan. One thing mentioned in #9042 and not yet mentioned in this issue is that another complication for reads before adding this feature was HWLB - which may decide to read from another node even if this one did have the data.

Right, I forgot about the other issue (as I usually do). I agree, this issue more concise and it lays out a clear plan on how to achieve the desired API.

**Comment by tzach** [comment]
We already have `LOCAL_ONE` as a Consistency Levels, refering to a local **DC**.
Can we find another term for this? at least LOCAL_NODE_ONLY

**Comment by bhalevy** [comment]
In addition to https://github.com/scylladb/scylladb/issues/16478#issuecomment-1865064081, I think it would simpler if the extension would be abstracted as yet another consistency level value rather than extending the query syntax.

**Comment by nyh** [comment]
@tzach @bhalevy as I explained in in the past (e.g., #9042) while in theory a consistency level sounds like a great idea, it's a problem because consistency levels aren't generic strings - they are sent in the protocol as numbers, and the driver needs to be modified to support each new one. We can easily monkey-patch the Python driver to add another consistency level, but we wouldn't want to start adding this feature to all other drivers...
On the other hand, extending the CQL syntax is easy and doesn't require driver changes.

**Comment by denesb** [comment]
There are further considerations, against the consistency-level argument:
* Setting the consistency level takes some fiddling. E.g. in CQLSH, it is a global setting. It is easy to make mistakes and think your node has all the data you think it should, while you accidentally read with another CL.
* Adding a new consistency level would mean carefully patching read and write paths in `storage_proxy.cc`, any rarely used path we'd miss, would lead to inaccurate tests. With a statement extension, we can derail the entire statement at a  very early point, having full control over execution. E.g. we can bypass `storage_proxy.cc` altogether and contact `database.cc` directly.

**Comment by bhalevy** [comment]
I concur, CQL extension it is then.

**Comment by tzach** [comment]
@nyh @denesb my comment is only about overloading the term "LOCAL", not about adding CLs.

**Comment by nyh** [comment]
I believe that pull request https://github.com/scylladb/scylladb/pull/24819 by @gusev-p, which will probably be merged soon, brings us very close to implementing this feature.

**Comment by denesb** [comment]
With [webshell](https://github.com/scylladb/scylladb/pull/24669), we can easily add a session option, which sets this flag for all writes, without having to add a CQL extension just for this.
